x/programs/test/programs/token/src/lib.rs

```diff
@@ -0,0 +1,199 @@
+
+    program
+        .state()
+        .store(StateKey::Balance(recipient), &(balance + amount))
+        .expect("failed to store balance");
+
+    true
+}
+
+/// Burn the token from the recipient.
+#[public]
+pub fn burn(context: Context<StateKey>, recipient: Address) -> u64 {
+    let Context { program, actor } = context;
+    owner_check(&program, actor);
+    program
+        .state()
+        .delete::<u64>(StateKey::Balance(recipient))
+        .expect("failed to burn recipient tokens")
+        .expect("recipient balance not found")
+}
+
+/// Gets the balance of the recipient.
+#[public]
+pub fn balance_of(context: Context<StateKey>, account: Address) -> u64 {
+    let Context { program, .. } = context;
+    program
+        .state()
+        .get(StateKey::Balance(account))
+        .expect("failure")
+        .unwrap_or_default()
+}
+
+#[public]
+pub fn allowance(context: Context<StateKey>, owner: Address, spender: Address) -> u64 {
+    let Context { program, .. } = context;
+    inner_allowance(&program, owner, spender)
+}
+
+pub fn inner_allowance(program:&Program<StateKey>, owner: Address, spender: Address) -> u64 {
+    program
+        .state()
+        .get::<u64>(StateKey::Allowance(owner, spender))
+        .expect("failure")
+        .unwrap_or_default()
+}
+
+#[public]
+pub fn approve(context: Context<StateKey>, spender: Address, amount: u64) -> bool {
+    let Context { program, actor } = context;
+    assert_ne!(actor, spender, "actor and spender must be different");
+    program
+        .state()
+        .store(StateKey::Allowance(actor,spender), &amount)
+        .expect("failed to store allowance");
+    true
+}
+
+/// Transfers balance from the sender to the the recipient.
+#[public]
+pub fn transfer(context: Context<StateKey>, recipient: Address, amount: u64) -> bool {
+    let Context { program, actor } = context;
+    assert_ne!(actor, recipient, "sender and recipient must be different");
+
+    // ensure the sender has adequate balance
+    let sender_balance = program
+        .state()
+        .get::<u64>(StateKey::Balance(actor))
+        .expect("failed to get sender balance")
+        .unwrap_or_default();
+
+    assert!(sender_balance >= amount, "invalid input");
+
+    let recipient_balance = program
+        .state()
+        .get::<u64>(StateKey::Balance(recipient))
+        .expect("failed to store recipient balance")
+        .unwrap_or_default();
+
+    // update balances
+    program
+        .state()
+        .store(StateKey::Balance(actor), &(sender_balance - amount))
+        .expect("failed to store sender balance");
+
+    program
+        .state()
+        .store(StateKey::Balance(recipient), &(recipient_balance + amount))
+        .expect("failed to store recipient balance");
+
+    true
+}
+
+#[public]
+pub fn transfer_from(
+    context: Context<StateKey>,
+    sender: Address,
+    recipient: Address,
+    amount: u64,
+) -> bool {
+    let Context {program, actor } = context;
```

Comment:
This is currently broken due to reuse of the context across the various functions.  I could create "inner" functions that do the actual allowance and transfer logic that instead take in the deconstructed pieces of the context, but that seems clunky.  Is there a cleaner way to handle this?